pkg/testing/runner/runner.go

```diff
@@ -237,64 +308,77 @@ func (r *Runner) runMachines(ctx context.Context, sshAuth ssh.AuthMethod, repoAr
+	sshPrivateKeyPath, err := filepath.Abs(filepath.Join(".integration-cache", "id_rsa"))
 	if err != nil {
 		return OSRunnerResult{}, fmt.Errorf("failed to determine OGC SSH private key path: %w", err)
 	}
 
-	logger.Logf("Starting SSH; connect with `ssh -i %s %s@%s`", sshPrivateKeyPath, machine.Layout.Username, machine.PublicIP)
+	logger.Logf("Starting SSH; connect with `ssh -i %s %s@%s`", sshPrivateKeyPath, instance.Username, instance.IP)
 	connectCtx, connectCancel := context.WithTimeout(ctx, 10*time.Minute)
 	defer connectCancel()
-	client, err := sshConnect(connectCtx, machine.PublicIP, machine.Layout.Username, sshAuth)
+	client, err := sshConnect(connectCtx, instance.IP, instance.Username, sshAuth)
 	if err != nil {
-		logger.Logf("Failed to connect to instance %s: %s", machine.PublicIP, err)
-		return OSRunnerResult{}, fmt.Errorf("failed to connect to instance %s: %w", machine.InstanceName, err)
+		logger.Logf("Failed to connect to instance %s: %s", instance.IP, err)
+		return OSRunnerResult{}, fmt.Errorf("failed to connect to instance %s: %w", instance.Name, err)
 	}
 	defer client.Close()
 	logger.Logf("Connected over SSH")
 
-	// prepare the host to run the tests
-	logger.Logf("Preparing instance")
-	err = batch.LayoutOS.Runner.Prepare(ctx, client, logger, batch.LayoutOS.OS.Arch, r.cfg.GOVersion, repoArchive, r.getBuildPath(batch))
+	if !instance.Prepared {
+		// prepare the host to run the tests
+		logger.Logf("Preparing instance")
+		err = batch.OS.Runner.Prepare(ctx, client, logger, batch.OS.Arch, r.cfg.GOVersion)
+		if err != nil {
+			logger.Logf("Failed to prepare instance: %s", err)
+			return OSRunnerResult{}, fmt.Errorf("failed to prepare instance %s: %w", instance.Name, err)
+		}
+
+		// now its prepared, add to state
+		instance.Prepared = true
+		err = r.addOrUpdateInstance(instance)
+		if err != nil {
+			return OSRunnerResult{}, fmt.Errorf("failed to save instance state %s: %w", instance.Name, err)
+		}
+	}
+
+	// copy the required files (done every run)
+	err = batch.OS.Runner.Copy(ctx, client, logger, repoArchive, r.getBuild(batch))
 	if err != nil {
-		logger.Logf("Failed to prepare instance: %s", err)
-		return OSRunnerResult{}, fmt.Errorf("failed to prepare instance %s: %w", machine.InstanceName, err)
+		logger.Logf("Failed to copy files instance: %s", err)
+		return OSRunnerResult{}, fmt.Errorf("failed to copy files to instance %s: %w", instance.Name, err)
 	}
 
 	// ensure that we have all the requirements for the stack if required
 	env := map[string]string{}
 	if batch.Batch.Stack != nil {
-		ch, err := r.getCloudForBatchID(batch.ID)
-		if err != nil {
-			return OSRunnerResult{}, err
+		// wait for the stack to be ready before continuing
+		r.stacksReady.Wait()
+		if r.stacksErr != nil {
+			return OSRunnerResult{}, fmt.Errorf("failed to save instance state %s: %w", instance.Name, err)
```

Comment:
typo, should be:
```
return OSRunnerResult{}, fmt.Errorf("failed to save instance state %s: %w", instance.Name, r.stacksErr)
```

Reply:
The error message itself is also wrong. I have updated, thanks for catching that!